### ISSUE TITLE
Toggle dark with @media (prefers-color-scheme: dark) such like ThemeProvider

### ISSUE THREAD

**Issue by Innei** [ant-design/ant-design#23371]
- [ ] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### What problem does this feature solve?
The theme color changes according to the system theme

主题色根据系统主题变换

### What does the proposed API look like?
```js
import 'antd/dist/antd.dark.css'
```
加载这个样式后，并不是直接应用到全局，而是可以通过一个类名，比如 `<body class="dark">`来切换主题。

再或者是把 less 颜色变量暴露到 css var
```css
:root {
    --red: #ff6347;
    --yellow: #ffc107;
    --blue: #6495ed;
}
@media screen and (prefers-color-scheme: dark) {
:root {
    color: #ddd;
    --red: #ad1a00;
    --blue: #1346a4;
  }
}
```

**Comment by Innei** [comment]
Similar #16661

**Comment by ntsd** [comment]
I'm doing this in .less file and it works fine.

```
@media (prefers-color-scheme: dark) {
  @import "~antd/dist/antd.dark.less";
}
```

for toggle dark mode I do this but get a little color bug.

```
.dark-theme {
  @import "~antd/dist/antd.dark.less";
}
```

**Comment by Innei** [comment]
I only used scss and css, and is there a way to use it in scss？

**Comment by ntsd** [comment]
> I only used scss and css, and is there a way to use it in scss？

I'm not sure but you can try it

```
@media (prefers-color-scheme: dark) {
  @import "~antd/dist/antd.dark.css";
}
```

**Comment by Innei** [comment]
> > I only used scss and css, and is there a way to use it in scss？
> 
> I'm not sure but you can try it
> 
> ```
> @media (prefers-color-scheme: dark) {
>   @import "~antd/dist/antd.dark";
> }
> ```
> 
> Just look at the `node_modules/antd/dist/` directory. Compiled styles is here.

I had to try, but failed. If do this, It will not bundle the `antd.dark` css file.

**Comment by dukesx** [comment]
> > > I only used scss and css, and is there a way to use it in scss？
> > 
> > 
> > I'm not sure but you can try it
> > ```
> > @media (prefers-color-scheme: dark) {
> >   @import "~antd/dist/antd.dark";
> > }
> > ```
> > 
> > 
> > Just look at the `node_modules/antd/dist/` directory. Compiled styles is here.
> 
> I had to try, but failed. If do this, It will not bundle the `antd.dark` css file.

Have you found a solution to import the file in SASS ?

**Comment by Innei** [comment]
> > > > I only used scss and css, and is there a way to use it in scss？
> > > 
> > > 
> > > I'm not sure but you can try it
> > > ```
> > > @media (prefers-color-scheme: dark) {
> > >   @import "~antd/dist/antd.dark";
> > > }
> > > ```
> > > 
> > > 
> > > Just look at the `node_modules/antd/dist/` directory. Compiled styles is here.
> > 
> > 
> > I had to try, but failed. If do this, It will not bundle the `antd.dark` css file.
> 
> Have you found a solution to import the file in SASS ?

Sorry, I don't have a good idea. I override all colors which antd used in my scss.

**Comment by dukesx** [comment]
I found a temporary and easy solution, I copied the css files of antd and antd dark and created two seperate scss file, each had a prefers color scheme only. So in prefers color scheme dark, I put all the dark css and in light the default css. So now each time I change my windows theme to light or dark, the website changes color flawlessly without any issue or extra work of modifying Less. 👍

**Comment by yoyo837** [comment]
@dukesx How to ensure that it can still work normally when update the `antd` version?

**Comment by dukesx** [comment]
Look, this is a trade off you will have to make. Ant Design doesn't have a dark mode toggle and it is impossible to toggle it live. Until Ant Design team wakes up and finds out a way to implement dark mode the right way, this is the easiest and "working" way. You can write custom css but that's time taking and your color scheme probably won't fit the components correctly. There is another way that you use nodejs to watch antd folder, if changes happen, you copy all content inside antd.css and antd.dark.css to your new files so that you keep it updated.

**Comment by tolking** [comment]
proposed

``` less
@primary-color: #1890ff;
@primary-dark-color: #1890ff;

body {
  --color-primary: @primary-color;
}
body.dark {
  --color-primary: @primary-dark-color;
}
body.auto {
  --color-primary: @primary-color;
  @media screen and (prefers-color-scheme: dark) {
    --color-primary: @primary-dark-color;
  }
}

a {
  color: var(--color-primary);
}
```

Default light theme

Add `dark` className to body to switch to dark mode

``` html
<body class="dark">
```

Add `auto` className to body to follow the system

``` html
<body class="auto">
```

**Comment by PaulPCIO** [comment]
AntDesign should really look into a ThemeProvider, or extend the ConfigProvider, it is really annoying to be forced to use css vars to customise the theme (considering it's not supported on all browsers). 

a system like [@zeit-ui/react](https://react.zeit-ui.co/en-us/guide/themes) would work so much better for dynamic theming. Take my [example](https://github.com/ant-design/ant-design/issues/477#issuecomment-633221661) in this thread, we have multiple clients, with their own branding colors, then we have a dark and light mode as well. Many methods have been tested but it simply does not scale, _eg; injecting the less vars as suggested in the docs._

The only performant solution I found was to do a complete cssvars override of the antd components.

https://github.com/PaulPCIO/antd-themed-cssvars

@afc163 🙏

**Comment by Rocinante89** [comment]
> AntDesign should really look into a ThemeProvider, or extend the ConfigProvider, it is really annoying to be forced to use css vars to customise the theme (considering it's not supported on all browsers).

Extending the ConfigProvider would be ideal. We have a similar requirement to your example re light/dark/whitelabeling and i'd really rather not have to go down the scss route. Proper dynamic theming is really the last piece that's missing from Antd

**Comment by JoseRFelix** [comment]
I stumbled on this issue and spent three days trying to find a decent solution. Since converting all variables to CSS vars and using SCSS is out of my scope, I  had to find an alternative solution. Injecting styles with Less was not scalable since it killed my app's performance. What I first decided was to just do as the Ant team does for their docs: prefetch styles, use [theme-switcher](https://www.npmjs.com/package/theme-switcher), and change the CSS directly on the HTML. 

At first, I was very happy with how fast it was compared to solutions using `less.modifyVars`, however, I hit a block since a flickering occurred when loading and injecting the styles; also, styles that were not Antd conflicted because of the order. Therefore, to solve this I created a library similar to theme-switcher, but using React's API. It comes with some extra features like selecting injection point and adding stylesheet injection status (to display a loading to avoid the flickering). 

You can check the library here: [react-css-theme-switcher](https://github.com/JoseRFelix/react-css-theme-switcher). It currently has 100% code coverage and is open to suggestions and PRs.

Currently, I can customize the theme however I want and my app performance was not as impacted. It is a good solution for the meantime.

I'll be writing a blog post on how to get custom CSS stylesheets and implement react-css-theme-switcher. As soon as it is written, I will update this comment with the link.  

TL;DR: My solution to dynamic theming was to use [react-css-theme-switcher](https://github.com/JoseRFelix/react-css-theme-switcher) and swap CSS stylesheets with the theme I need within React.

**Comment by Rocinante89** [comment]
> I stumbled on this issue and spent three days trying to find a decent solution. Since converting all variables to CSS vars and using SCSS is out of my scope, I had to find an alternative solution. Injecting styles with Less was not scalable since it killed my app's performance. What I first decided was to just do as the Ant team does for their docs: prefetch styles, use [theme-switcher](https://www.npmjs.com/package/theme-switcher), and change the CSS directly on the HTML.
> 
> At first, I was very happy with how fast it was compared to solutions using `less.modifyVars`, however, I hit a block since a flickering occurred when loading and injecting the styles; also, styles that were not Antd conflicted because of the order. Therefore, to solve this I created a library similar to theme-switcher, but using React's API. It comes with some extra features like selecting injection point and adding stylesheet injection status (to display a loading to avoid the flickering).
> 
> You can check the library here: [react-css-theme-switcher](https://github.com/JoseRFelix/react-css-theme-switcher). It currently has 100% code coverage and is open to suggestions and PRs.
> 
> Currently, I can customize the theme however I want and my app performance was not as impacted. It is a good solution for the meantime.
> 
> I'll be writing a blog post on how to get custom CSS stylesheets and implement react-css-theme-switcher. As soon as it is written, I will update this comment with the link.
> 
> TL;DR: My solution to dynamic theming was to use [react-css-theme-switcher](https://github.com/JoseRFelix/react-css-theme-switcher) and swap CSS stylesheets with the theme I need within React.

Looks nice. Looking forward to reading it!

**Comment by marklai1998** [comment]
Same issue here, I use injectType: 'lazyStyleTag' to get around this.
But the drawback is quite significant, both of the CSS is at my bundle righty now without any tree shaking.

**Comment by nhevia** [comment]
This should really be a priority, runtime theming is a pretty much used feature nowadays. Spent several hours to make it work, using a workaround.

> The only performant solution I found was to do a complete cssvars override of the antd components.
> 
> https://github.com/PaulPCIO/antd-themed-cssvars

Really appreciate the work done here. Performant and extendible. Too bad that since it's not officially supported, something might break the theme in future releases (minor changes to antd, new components).

For anyone that wants to use the @PaulPCIO approach:

Usage:
1) git clone the repo and copy all files locally (node-sass is required for CRA)
2) Import index.scss in your App.js (or relevant) file. 
3) Create themes.scss at index level, import it at the bottom of index.
4) In themes.scss customize your themes:

```css
.light-theme {
  --primary-color: #007bff;
  --disabled-color: #6c757d;
  --bodyBackground: #ffffff;
  // ... etc
}

.dark-theme {
  --primary-color: #000000;
  --disabled-color: #4a4a4b;
  --bodyBackground: #000000;
  // ... etc
}
```

5) Assign light/dark className to App, change on button event click.

**Comment by saostad** [comment]
+1 to ThemeProvider

**Comment by yisiqi** [comment]
I just created a PR to support automatic dark mode through theme generation.

You can use the `umi-plugin-antd-theme` plugin to generate an automatic dark mode theme.

chenshuai2144/antd-pro-merge-less/pull/24

**Comment by Rocinante89** [comment]
@yisiqi, will that work for those of use using Create React App or Next.js?

**Comment by IronSean** [comment]
CSS Variable support is pretty widespread these days, if the less variables could all just be replaced with CSS Variables I think it would simplify a lot of use cases. Not to mention CRA and Storybook among others not playing nicely with LESS out of the box. I currently have a big file where I assign all antd css variables to less variables to help get around some of these issues.

**Comment by menosprezzi** [comment]
Hello people!

So why we have this awesome PR abandoned at all? #26139

I was starting to create a PR with something like it...

**Comment by emindeniz99** [comment]
I think that this is very good issue. Because css variables helps us to change theme at runtime. 
Few months ago, I need an UI framework to support color changes for multi tenant app. I could not find a good solution for antd to change colors at runtime.

**Comment by arifszn** [comment]
> I only used scss and css, and is there a way to use it in scss？

I used `sass:meta` in scss like this for dynamic dark mode:

```
@use "sass:meta";

body.light {
    background-color: #fff;
    @include meta.load-css("node_modules/antd/dist/antd.css", $with: null);
}

body.dark {
    background-color: #000;
    @include meta.load-css("node_modules/antd/dist/antd.dark.css", $with: null);
}
```

**Comment by stam** [comment]
It looks like ant design comes with an experimental ConfigProvider nowadays, which accepts a theme prop: https://ant.design/docs/react/customize-theme-variable

I however haven't found a way to export all nextjs dark theme variables, as they are only defined in deeply nested less files

**Comment by doong-jo** [comment]
```javascript
const isDarkTheme = window.matchMedia("(prefers-color-scheme: dark)").matches;
import(`antd/dist/antd.${isDarkTheme ? 'dark.' : ''} css`);
```

Write your root component. it works :)

**Comment by MarkLyck** [comment]
@afc163 any update on being able to toggle darkmode with ConfigProvider? 🙏

**Comment by ISBN9527** [comment]
> @media (prefers-color-scheme: dark) {
>   @import "~antd/dist/antd.dark.less";
> }

Does this need to set the class on the body? I tried it but it didn't work. I'm a little confused. Thank

**Comment by arifszn** [comment]
I used this to enable dynamic dark mode in a scss file and it's working:

```scss
@use 'sass:meta';

body:not(.dark) {
  @include meta.load-css('node_modules/antd/dist/antd.css', $with: null);
}

body.dark {
  @include meta.load-css('node_modules/antd/dist/antd.dark.css', $with: null);
}
```

The dark mode switch just adds a `dark` class to the body tag.

Demo: https://arifszn.github.io

**Comment by huacnlee** [comment]
Use SCSS Nesting import:

https://sass-lang.com/documentation/at-rules/import#nesting

Create a `antd.dark.scss`

```scss
@media (prefers-color-scheme: dark) {
  @import '~antd/dist/antd.dark';
}
```

Import `antd.dark.scss` in your JS `index.ts`

```ts
import React from 'react';

import './antd.less';
import './antd.dark.scss';
```

**Comment by arkhamRejek** [comment]
> I used this to enable dynamic dark mode in a scss file and it's working:
> 
> ```css-scss
> @use 'sass:meta';
> 
> body:not(.dark) {
>   @include meta.load-css('node_modules/antd/dist/antd.css', $with: null);
> }
> 
> body.dark {
>   @include meta.load-css('node_modules/antd/dist/antd.dark.css', $with: null);
> }
> ```
> 
> The dark mode switch just adds a `dark` class to the body tag.
> 
> Demo: https://arifszn.github.io

@arifszn  your solution works! Cheers

Anyone else trying to use this solution 
 
* set your .less variables as normal 
* compile your dark & light files (the command below will handle that just update your paths) 
```json
    "ant:theme": "lessc --js public/styles/ant/ant.less public/styles/ant-light.css && lessc --js public/styles/ant/ant-dark.less public/styles/ant-dark.css"
```
* create a [whatever you want to name it].scss file and import your recently compiled css using the method above. In my 
```scss
@use 'sass:meta';

html:not(.dark) {
  @include meta.load-css('./ant-light.css', $with: null);
}

html.dark {
  @include meta.load-css('./ant-dark.css', $with: null);
}
```

I did html because i'm using https://github.com/pacocoursey/next-themes 

understand what sass:meta is doing https://sass-lang.com/documentation/modules/meta

**Comment by wellitongervickas** [comment]
It should help someone that need to handle less vars to switch between themes
https://www.npmjs.com/package/antd-theme-vars

**Comment by YusukeSano** [comment]
What is the reason for the current lack of attention to this issue?

**Comment by long76** [comment]
antd 6 solution
```tsx
<ConfigProvider
         theme={{
            algorithm: window.matchMedia('(prefers-color-scheme: dark)').matches
              ? theme.darkAlgorithm
              : theme.defaultAlgorithm
         }}
       >
    <App>
         <Layout style={{height: '100vh'}}>
         </Layout>
    </App>
</ConfigProvider>
```
style need for layout for black background, on `Space`, `Grid` and etc. background don't apply
